src/engraving/libmscore/slur.cpp

```diff
@@ -1741,56 +1723,75 @@ SpannerSegment* Slur::layoutSystem(System* system)
     return slurSegment;
 }
 
-void Slur::computeUp()
+void Slur::computeUp(SlurSegment* slurSegment)
 {
-    //
-    // assumption:
-    // slurs have only chords or rests as start/end elements
-    //
-    if (startCR() == 0 || endCR() == 0) {
-        _up = true;
-        return;
-    }
-    Chord* c1 = startCR()->isChord() ? toChord(startCR()) : 0;
-    Chord* c2 = endCR()->isChord() ? toChord(endCR()) : 0;
-    if (c2 && startCR()->measure()->system() != endCR()->measure()->system()) {
-        // If the end chord is in a different system its direction may
-        // have never been computed, so we need to compute it here.
-        c2->computeUp();
-    }
-
-    if (c1 && c1->beam() && c1->beam()->cross()) {
-        // TODO: stem direction is not finalized, so we cannot use it here
+    switch (_slurDirection) {
+    case DirectionV::UP:
         _up = true;
-        return;
-    }
+        break;
+    case DirectionV::DOWN:
+        _up = false;
+        break;
+    case DirectionV::AUTO:
+    {
+        //
+        // assumption:
+        // slurs have only chords or rests as start/end elements
+        //
+        if (startCR() == 0 || endCR() == 0) {
+            _up = true;
+            break;
+        }
+        Chord* c1 = startCR()->isChord() ? toChord(startCR()) : 0;
+        Chord* c2 = endCR()->isChord() ? toChord(endCR()) : 0;
+        if (c2 && startCR()->measure()->system() != endCR()->measure()->system()) {
+            // If the end chord is in a different system its direction may
+            // have never been computed, so we need to compute it here.
+            c2->computeUp();
+        }
 
-    _up = !(startCR()->up());
+        if (slurSegment && _sourceStemArrangement != -1 && _sourceStemArrangement != calcStemArrangement(c1, c2)) {
+            // copy & paste from incompatible stem arrangement, so reset bezier points
+            for (int g = 0; g < (int)Grip::GRIPS; ++g) {
+                slurSegment->ups((Grip)g) = UP();
+            }
+        }
```

Comment:
I'm doubtful about having this in `computeUp()`. It looks like we should be able to do `computeUp()` once for the whole slur ("once" as opposed to "for every segment"), and move this segment-dependent logic back to `layoutSystem`. Or am I missing something?

Reply:
`I think it's possible to move that segment-related part out of this method though`

do it

Reply:
done!